### PR TITLE
Fix usb mass storage device issues

### LINKED ISSUE #77175: SCSI Read Capacity reports wrong capacity when used with USB Mass storage device
**Describe the bug**
When a device is configured as a USB Mass Storage device, is sometimes report a wrong capacity to the host computer through
the SCSI Read Capacity command.

**To Reproduce**
1. Configure a development board using STM32U585 CPU with USB Mass Storage function
2. Connect that development kit to a host computer running Linux
3. The computer may not see the device, or see it with a capacity that does not match what is expected (see `dmesg` output)
4. If you are using a Linux computer, the following command will show the reported capacity of the mass storage device: 'sudo sg_readcap /dev/sdX'. You should be able to see the wrong capacity reported there.

**Expected behavior**
We expect the host computer to obtain valid capacity values for USB mass storages devices that are connected

**Impact**
It renders the USB mass storage litteraly unusable

**Logs**
Dmesg output on the host computer: 

```
[61240.809883] sd 2:0:0:0: Attached scsi generic sg1 type 0
[61240.810439] sd 2:0:0:0: [sdb] Unsupported sector size 263520264.
[61240.812816] sd 2:0:0:0: [sdb] 0 512-byte logical blocks: (0 B/0 B)
[61240.812825] sd 2:0:0:0: [sdb] 263520264-byte physical blocks
[61240.813034] sd 2:0:0:0: [sdb] Write Protect is off
[61240.813040] sd 2:0:0:0: [sdb] Mode Sense: 0f b5 00 08
[61240.813250] sd 2:0:0:0: [sdb] No Caching mode page found
[61240.813256] sd 2:0:0:0: [sdb] Assuming drive cache: write through
[61240.813645] sd 2:0:0:0: [sdb] Attached SCSI removable disk
```

**Environment**

 - OS: Linux
 - Zephyr SDK
 - Commit SHA: 0d56a3c3f4c333b08b0fecde02a9c032ee6c2d2f

### PR DESCRIPTION
**Pull request by ghost**
disk: disk_access: changed mutex for a spinlock

disk_access functions were called from an interupt context, thus
crashing on mutexes (no sleep is allowed in interrupt context).
The usage of a spinlock is permitted. The functions guarded by
the spinlock are executed very quickly anyway, so the spinlock
is applicable.

Stack trace from VSCode when the crash occured:

image

Log output from Zephyr: 
```
ASSERTION FAIL [!arch_is_in_isr()] @ WEST_TOPDIR/zephyr-deps/zephyr/kernel/mutex.c:111
        mutexes cannot be used inside ISRs
[00:00:18.066,000] <err> os: r0/a1:  0x00000004  r1/a2:  0x0000006f  r2/a3:  0x00000001
[00:00:18.066,000] <err> os: r3/a4:  0x00000004 r12/ip:  0x0002c1b9 r14/lr:  0x08011ead
[00:00:18.066,000] <err> os:  xpsr:  0x01000059
[00:00:18.066,000] <err> os: Faulting instruction address (r15/pc): 0x0802a468
[00:00:18.066,000] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
[00:00:18.066,000] <err> os: Fault during interrupt handling

[00:00:18.066,000] <err> os: Current thread: 0x20001b18 (idle)
[00:00:18.128,000] <err> os: Halting system
```

usb: fix wrong capacity report for USB mass storage device
    
Usb "write" method never copies the input data buffer, it uses the
buffer pointer directly. The fact that it was on the stack made
the buffer ephemeral, and could disappear before the USB transfer
actually occurred. The data transmitted was then random. This commit
converts the transfer buffer to a static buffer, so that it always
exists for the duration of the USB transfer.

Known side-effect: if "read capacity" command is read twice
simultaneously, the same buffer will be used twice. As the capacity of a
USB drive should not change between 2 calls, this side effect can be
ignored.

Signed-off-by: Alexi Demers <alexi.demers@axceta.com>

Fixes #77175 and #77176

### CODE REVIEW

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: subsys/disk/disk_access.c [hunk 1]
```diff
@@ -58,8 +58,7 @@ int disk_access_init(const char *pdrv)
 	struct disk_info *disk = disk_access_get_di(pdrv);
 	int rc = -EINVAL;
 
-	if ((disk != NULL) && (disk->ops != NULL) &&
-				(disk->ops->init != NULL)) {
+	if ((disk != NULL) && (disk->ops != NULL) && (disk->ops->init != NULL)) {
```
[the remark this answers is not in the document]
Comment (ademers-axceta) [reply subsys/disk/disk_access.c]: Sorry, I reverted those.

File: subsys/disk/disk_access.c [hunk 2]
```diff
@@ -21,15 +21,15 @@ LOG_MODULE_REGISTER(disk);
 static sys_dlist_t disk_access_list = SYS_DLIST_STATIC_INIT(&disk_access_list);
 
 /* lock to protect storage layer registration */
-static K_MUTEX_DEFINE(mutex);
+static struct k_spinlock lock;
```
[the remark this answers is not in the document]
Comment (ademers-axceta) [reply subsys/disk/disk_access.c:24]: I added a stack trace with line information and the Zephyr log output related to the crash. Let me know if this is enough

**Comment by danieldegrasse** [comment]
So to be clear, the issue is that `mass_storage_bulk_out` is being called from an ISR context? @jfischer-no, should the USB device drivers be calling this from an ISR context? Looking at the Kinetis USB driver, it seems that call is made from a thread context (an ISR notifies the thread)

**Comment by jfischer-no** [comment]
> So to be clear, the issue is that `mass_storage_bulk_out` is being called from an ISR context? @jfischer-no, should the USB device drivers be calling this from an ISR context? Looking at the Kinetis USB driver, it seems that call is made from a thread context (an ISR notifies the thread)

Yes, this is a known issue in legacy USB device support, most drivers are fixed not to call callbacks from ISR context. Btw, STM32 shim driver has another mutex issue, https://github.com/zephyrproject-rtos/zephyr/issues/61464.

File: subsys/disk/disk_access.c [hunk 3]
```diff
@@ -21,15 +21,15 @@ LOG_MODULE_REGISTER(disk);
...
 
 struct disk_info *disk_access_get_di(const char *name)
 {
 	struct disk_info *disk = NULL, *itr;
 	size_t name_len = strlen(name);
 	sys_dnode_t *node;
 
-	k_mutex_lock(&mutex, K_FOREVER);
+	k_spinlock_key_t spinlock_key = k_spin_lock(&lock);
```
Comment (jfischer-no) [inline subsys/disk/disk_access.c:32]: k_spinlock_key_t definition should be at the top of the block, like other variables.

File: subsys/disk/disk_access.c [hunk 4]
```diff
@@ -124,47 +124,52 @@ int disk_access_register(struct disk_info *disk)
...
 	int rc = 0;
 
-	k_mutex_lock(&mutex, K_FOREVER);
+	k_spinlock_key_t spinlock_key = k_spin_lock(&lock);
 	if ((disk == NULL) || (disk->name == NULL)) {
 		LOG_ERR("invalid disk interface!!");
 		rc = -EINVAL;
 		goto reg_err;
 	}
+	k_spin_unlock(&lock, spinlock_key);
```
Comment (jfischer-no) [inline subsys/disk/disk_access.c:133]: Looks like there is no need for a lock and you can replace `goto` with `return -EINVAL`, and you can remove rc variable. Only sys_dlist_append() needs to be protected, same applies for the  disk_access_unregister().

File: subsys/disk/disk_access.c [hunk 5]
```diff
@@ -124,47 +124,52 @@ int disk_access_register(struct disk_info *disk)
...
 
 	if (disk_access_get_di(disk->name) != NULL) {
 		LOG_ERR("disk interface already registered!!");
 		rc = -EINVAL;
-		goto reg_err;
+		return rc;
```
Comment (jfischer-no) [inline subsys/disk/disk_access.c:138]: just `return -EINVAL` should be fine.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Comment by henrikbrixandersen** [comment]
@ademers-axceta How come this PR is targeting `v3.6-branch`?

**Review by tmon-nordic — DISMISSED** [review]
The USB MSC class change is probably as good as it is reasonably to have on code that is scheduled for deprecation and removal. The new MSC implementation does not share this issue (and a bunch of others).

**Review by henrikbrixandersen — DISMISSED** [review]
Blocking due to the target branch being v3.6-branch pending further explanation.

**Comment by ademers-axceta** [comment]
> @ademers-axceta How come this PR is targeting `v3.6-branch`?

The bug was encountered on version 3.6. I am suggesting to integrate in in 3.6, as it is still supported as far as I know? I can port it to 3.7 and main branches afterward if it applies?

If you prefer having it on 3.7 or main only, I will adjust.

**Comment by henrikbrixandersen** [comment]
> > @ademers-axceta How come this PR is targeting `v3.6-branch`?
> 
> The bug was encountered on version 3.6. I am suggesting to integrate it in 3.6, as it is still supported as far as I know? I can port it to 3.7 and main branches afterward if it applies?
> 
> If you prefer having it on 3.7 or main only, I will adjust.

The process for this would be to get the fix into `main` first and then backport it to the relevant, active releases.
Please change the target branch to `main` and I'll make the necessary label additions for automated backports to `v3.7-branch` and `v3.6-branch` (these require separate reviews and approvals).

**Comment by ademers-axceta** [comment]
> k_spinlock_key_t spinlock_key = k_spin_lock(&lock)

Ah ok! I was not aware of that backporting method. I will adjust the target branch, thanks

**Comment by ademers-axceta** [comment]
> > > @ademers-axceta How come this PR is targeting `v3.6-branch`?
> > 
> > 
> > The bug was encountered on version 3.6. I am suggesting to integrate it in 3.6, as it is still supported as far as I know? I can port it to 3.7 and main branches afterward if it applies?
> > If you prefer having it on 3.7 or main only, I will adjust.
> 
> The process for this would be to get the fix into `main` first and then backport it to the relevant, active releases. Please change the target branch to `main` and I'll make the necessary label additions for automated backports to `v3.7-branch` and `v3.6-branch` (these require separate reviews and approvals).

Ah ok! I was not aware of that backporting method. I will adjust the target branch, thanks

**Comment by henrikbrixandersen** [comment]
> Ah ok! I was not aware of that backporting method. I will adjust the target branch, thanks

You will also need to fix the commit logs by removing `subsys: ´ from the title in order to satisfy the compliance check.

**Comment by danieldegrasse** [comment]
> Yes, this is a known issue in legacy USB device support, most drivers are fixed not to call callbacks from ISR context. Btw, STM32 shim driver has another mutex issue, https://github.com/zephyrproject-rtos/zephyr/issues/61464.

So I guess my question is then why don't we fix the STM32 driver, instead of modifying the subsystem functions to be ISR-safe?

**Comment by ademers-axceta** [comment]
> > Yes, this is a known issue in legacy USB device support, most drivers are fixed not to call callbacks from ISR context. Btw, STM32 shim driver has another mutex issue, #61464.
> 
> So I guess my question is then why don't we fix the STM32 driver, instead of modifying the subsystem functions to be ISR-safe?

I would argue that making the functions ISR-safe is a more robust way to fix the issue, and it will probably fix other issues in other drivers. If the mutex was absolutely necessary, then the fix would have been to adjust the STM32 driver, but as the protected code executes very quickly, the spinlock seems to me to be a better option anyway

**Comment by danieldegrasse** [comment]
> I would argue that making the functions ISR-safe is a more robust way to fix the issue, and it will probably fix other issues in other drivers. If the mutex was absolutely necessary, then the fix would have been to adjust the STM32 driver, but as the protected code executes very quickly, the spinlock seems to me to be a better option anyway.

I agree that it is certainly a simpler solution. I guess what I'm trying to understand is if the STM32 driver should be making these calls from an ISR, or if the `memoryWrite` function is only intended to be called from a thread content. Looking at the implementation, it seems to signal a thread that memory is ready to be written to disk, so I suspect the function is intended to be called from an ISR, I just want to confirm. Perhaps @jfischer-no knows the answer here?

> but as the protected code executes very quickly, the spinlock seems to me to be a better option anyway

This is *generally* true, but there isn't a constraint here. For example, the SDMMC disk driver checks card presence when getting disk status. For some card detection methods, the host needs to wait on the SD device to confirm it is present (which is a blocking operation). I'm still not opposed to the spinlock solution, I just want to make sure this is the best option, since when the choice is between changing a vendor driver or changing generic code I tend to bias towards modifying the driver.

**Comment by tmon-nordic** [comment]
> I agree that it is certainly a simpler solution. I guess what I'm trying to understand is if the STM32 driver should be making these calls from an ISR, or if the `memoryWrite` function is only intended to be called from a thread content. Looking at the implementation, it seems to signal a thread that memory is ready to be written to disk, so I suspect the function is intended to be called from an ISR, I just want to confirm. Perhaps @jfischer-no knows the answer here?

This MSC implementation is essentially flawed because it uses global state variables without any proper synchronization. The device-next MSC implementation is significantly better because it correctly sequences operations (always in thread context) and also properly reports errors (e.g. disk removed).

If the disk subsystem works correctly with the device-next MSC, then I wouldn't change the disk subsystem.

**Comment by jfischer-no** [comment]
> I agree that it is certainly a simpler solution. I guess what I'm trying to understand is if the STM32 driver should be making these calls from an ISR, or if the `memoryWrite` function is only intended to be called from a thread content. Looking at the implementation, it seems to signal a thread that memory is ready to be written to disk, so I suspect the function is intended to be called from an ISR, I just want to confirm. Perhaps @jfischer-no knows the answer here?
> 
> > but as the protected code executes very quickly, the spinlock seems to me to be a better option anyway
> 
> This is _generally_ true, but there isn't a constraint here. For example, the SDMMC disk driver checks card presence when getting disk status. For some card detection methods, the host needs to wait on the SD device to confirm it is present (which is a blocking operation). I'm still not opposed to the spinlock solution, I just want to make sure this is the best option, since when the choice is between changing a vendor driver or changing generic code I tend to bias towards modifying the driver.

The mutex there only protects the dlist (register/unregister), not accesses to the disk device (IIRC there are some in the drivers as well), so I think the changes are fine. Obviously, the second commit does not fix a bug and cannot be backported, but the first commit does.

**Comment by danieldegrasse** [comment]
> The mutex there only protects the dlist (register/unregister), not accesses to the disk device (IIRC there are some in the drivers as well), so I think the changes are fine. Obviously, the second commit does not fix a bug and cannot be backported, but the first commit does.

This is true- at the end of the day, this isn't really *that* large of a change to common code, and it is pretty low risk.

**Review by danieldegrasse — DISMISSED** [review]

**Comment by ademers-axceta** [comment]
@jfischer-no are you okay with the adjustements I made based on your comments? I need your approval to be able to merge :) (changes requested)

File: subsys/disk/disk_access.c [hunk 6]
```diff
@@ -167,52 +167,47 @@ int disk_access_ioctl(const char *pdrv, uint8_t cmd, void *buf)
...
 	}
+
+	spinlock_key = k_spin_lock(&lock);
 	/* remove disk node from the list */
 	sys_dlist_remove(&disk->node);
 	LOG_DBG("disk interface(%s) unregistered", disk->name);
-unreg_err:
-	k_mutex_unlock(&mutex);
-	return rc;
+	k_spin_unlock(&lock, spinlock_key);
```
Comment (ydamigos) [inline subsys/disk/disk_access.c:211]: Just a nitpick, we could move it above the LOG_DBG.

**Review by jfischer-no — DISMISSED** [review]
Thanks

File: subsys/disk/disk_access.c (hunk 6 above)
Comment (ademers-axceta) [reply subsys/disk/disk_access.c:210]: Done

**Review by jfischer-no — APPROVED** [review]

**Review by ydamigos — APPROVED** [review]

**Review by danieldegrasse — APPROVED** [review]

**Pull request merged**